{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create vcf file with all possible peptide elongations resulting from a SNV in a STOP codon\n",
    "1. Get the nucleotide sequences of all canonical peptides\n",
    "    - Downloaded Coding Sequences (cds) from ensembl (https://ftp.ensembl.org/pub/release-112/fasta/homo_sapiens/).\n",
    "\n",
    "    Function *parse_cds_to_dict()*\n",
    "    - 1.0 Check if gene_symbol already processed previously. If yes, skip to next.\n",
    "    - 1.1 Find first ATG in cds, if not move to next and print to std.out\n",
    "    - 1.2 Check for downstream STOP codon, if not print std.out\n",
    "    - 1.3 If ATG and STOP found, store gene_symbol, ENSG, ENST and description in df\n",
    "    - 1.4 Store gene_symbol and CDS in seq object\n",
    "    - 1.5 Mitochondrial genes will be dropped as they follow diffferent rules of translationa nd transcription and often have no UTRs.\n",
    "\n",
    "    After this step, we obtain 19 360 CDS sequences of unique genes that have a \"valid sequence\"\n",
    "\n",
    "2. Get cDNA with ENST-ID of CDS\n",
    "    - 2.1 Create a cdna_dict from cDNA ensmbl file with ENST_ID:cDNA_seq (function *build_cdna_dict()*)\n",
    "    Function *extract_enstID()* to get transcript ID of CDS entry\n",
    "    Function *create_enst_dict()* creates dictionary with gene:ENST_ID\n",
    "    - 2.2 Function *add_cdna_to_cds()* to match CDS with cDNA using the transcript ID (ENST)\n",
    "    \n",
    "    For 892 of the 19 360 CDS, we found no matching ENST-ID in the cDNA.\n",
    "\n",
    "3. Mutating the STOP codon of the 18 467 CDS\n",
    "    - 3.1 Function *trim_cdna_to_overlap()* will keep only the 3'UTR of the cDNA (seq after end of CDS).\n",
    "    - 3.2 Function *find_ds_STOP_in3UTR()* checks for downstream STOP codons in the 3'UTR. If none found, the entry will be discarded.\n",
    "    Out of 18468, we removed 300 genes due to missing dsSTOP (1.62%).\n",
    "    - 3.3 Function *generate_mutated_sequences_with_utr()* introduces all possible non-silent mutations in the original STOP codon of the CDS. \n",
    "    The possibile mutated codons are stored in nonsilentSNV_dict.\n",
    "    - 3.4 Function *translate_mutated_cds_to_pep()* uses Bio.Seq object to translate the elongtaed CDS to peptide\n",
    "\n",
    "4. Write fasta files\n",
    "    - 4.1 Function *update_header()* will create a unique ID for each mutated seq in the format >[gene_symbol]_n \n",
    "    - 4.2 Function *write_fasta_files()* creates 2 output files:\n",
    "        - peptide seq in amino acids\n",
    "        - mutated and elongated CDS in nucleotides\n",
    "\n",
    "## Output:\n",
    "The final two fasta files contain the exact same sequneces of the mutated sequences\n",
    "in amino acid (pep) and nucleotide (CDS) format.\n",
    "Mutated sequence: [8 last codons of CDS] + [mutated STOP codon] + [3'UTR to downstream STOP codon]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "wdir = '/users/genomics/lilian/STOPcodons_in_humangenes/STOPloss_canonicalPeptides/'\n",
    "outpath = wdir+'data/'\n",
    "\n",
    "old_stdout = sys.stdout\n",
    "log_file = open(wdir+'scripts/01_get_seq_elongation_v2.0.log', 'w')\n",
    "sys.stdout = log_file"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Genome files. ensembl version 112"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ref files\n",
    "GRCh38_cds_fasta = \"/data/genomics/lilian/Ref_genomes/GRCh38/Homo_sapiens.GRCh38.cds.all.fa\"\n",
    "GRCh38_cDNA_fasta = \"/data/genomics/lilian/Ref_genomes/GRCh38/Homo_sapiens.GRCh38.cdna.all.fa\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 1: Get the CDS for all unique genes\n",
    "\n",
    "One gene symbol can have several cds. We will keep the first one that starts with ATG and ends with one of the three common STOP codons.\n",
    "\n",
    "- Check if gene_symbol already processed previously. If yes, skip to next.\n",
    "- Check if transcript_biotype:protein_coding. If no, skip to next.\n",
    "- Keep only sequences:\n",
    "    1. ... starting with ATG\n",
    "    2. ... ending with a STOP codon\n",
    "    3. ... having STOP in-frame with START codon (seq_len = multiple of 3)\n",
    "- Check if termiates in STOP codon, if not print std.out\n",
    "- Store gene with full header and CDS in final dictionary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_cds_to_dict(cds_filepath):\n",
    "    \"\"\"\n",
    "    Extracting CDS for all unique genes into dictionary\n",
    "    Only considering transcripts with \"transcript_biotype:protein_coding\" in their header\n",
    "    Will take the first cds transcript matching \n",
    "    \"\"\"\n",
    "    unique_cds_dict = {}\n",
    "    current_gene = None\n",
    "    current_sequence = []\n",
    "    handled_genes = [] # list to store all genes with transcript biotype protein_coding\n",
    "\n",
    "    with open(cds_filepath, 'r') as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if \"transcript_biotype:protein_coding\" not in line:\n",
    "                    current_gene = None  # Skip this entry\n",
    "                    continue\n",
    "\n",
    "                # If current gene and sequence, store to dictionary\n",
    "                if current_gene is not None and current_gene not in unique_cds_dict:\n",
    "                    cds = ''.join(current_sequence)\n",
    "\n",
    "                    # add seq to dictionary 1.) IF starts with ATG and 2.) IF contains in-frame STOP \n",
    "                    start_index = cds.find('ATG')\n",
    "                    # If no ATG found, return False\n",
    "                    if start_index == -1:\n",
    "                        print(f\"Skipping entry for {current_gene}: sequence does not start with 'ATG'.\")\n",
    "\n",
    "                    # if ATG found, check for in-frame stop codons (TAG, TGA, TAA)\n",
    "                    no_stop_found = True\n",
    "                    for i in range(start_index, len(cds), 3):\n",
    "                        codon = cds[i:i+3]\n",
    "                        if codon in ['TAG', 'TGA', 'TAA']:\n",
    "                            no_stop_found = False\n",
    "                            #cds_start_stop = full_sequence[start_index:i+3]\n",
    "                            unique_cds_dict[current_gene] = {'cds_info': current_header, \n",
    "                                                             'cds_sequence': cds, \n",
    "                                                             'ATG_pos': start_index,\n",
    "                                                             'STOP_pos': i}\n",
    "                    if no_stop_found:\n",
    "                        print(f\"Skipping entry for {current_gene}: sequence misses expected STOP codon.\")\n",
    "\n",
    "                current_header = line\n",
    "                # get gene_symbol from header\n",
    "                header = line.split()\n",
    "                for part in header:\n",
    "                    if part.startswith(\"gene_symbol:\"):\n",
    "                        current_gene = part.split(\":\")[1].split('.')[0] \n",
    "                        handled_genes.append(current_gene)\n",
    "\n",
    "                # if gene_symbol is already in dictionary, skip lines\n",
    "                if current_gene in unique_cds_dict:\n",
    "                    current_gene = None  \n",
    "\n",
    "                # Reset the current sequence list for the new gene\n",
    "                current_sequence = []\n",
    "            else:\n",
    "                 # Only append sequence if we have a valid current_gene\n",
    "                if current_gene is not None:\n",
    "                    current_sequence.append(line)\n",
    "\n",
    "        # After finishing the loop, store the last gene and sequence\n",
    "        if current_gene is not None and current_gene not in unique_cds_dict:\n",
    "            cds = ''.join(current_sequence)\n",
    "\n",
    "            # add seq to dictionary 1.) IF starts with ATG and 2.) IF contains in-frame STOP \n",
    "            start_index = cds.find('ATG')\n",
    "            # If no ATG found, return False\n",
    "            if start_index == -1:\n",
    "                print(f\"Skipping entry for {current_gene}: sequence does not start with 'ATG'.\")\n",
    "\n",
    "            # if ATG found, check for in-frame stop codons (TAG, TGA, TAA)\n",
    "            no_stop_found = True\n",
    "            for i in range(start_index, len(cds), 3):\n",
    "                codon = cds[i:i+3]\n",
    "                if codon in ['TAG', 'TGA', 'TAA']:\n",
    "                    no_stop_found = False\n",
    "                    #cds_start_stop = full_sequence[start_index:i+3]\n",
    "                    unique_cds_dict[current_gene] = {'cds_info': current_header, \n",
    "                                                        'cds_sequence': cds, \n",
    "                                                        'ATG_pos': start_index,\n",
    "                                                        'STOP_pos': i}\n",
    "            if no_stop_found:\n",
    "                print(f\"Skipping entry for {current_gene}: sequence misses expected STOP codon.\")\n",
    "\n",
    "\n",
    "    return unique_cds_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Load CDS into dict ...\")\n",
    "cds_dict = parse_cds_to_dict(GRCh38_cds_fasta)\n",
    "print(f\"\\nNumber of unique genes with cds in dictionary:{len(cds_dict.keys())}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 2: Get cDNA, containing 3'UTR after STOP\n",
    "- from header_info in cds_dict, take the transcript_ID\n",
    "- grep transcript_ID in cDNA file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_enstID(header_info):\n",
    "    \"\"\"\n",
    "    Get ENST-ID from header line in format \n",
    "    '>ENST00000289823.1 cdna chromosome:NCBI35:8:21922367:21927699:1 gene:ENSG00000158815.1 gene_biotype:protein_coding transcript_biotype:protein_coding'\n",
    "    \"\"\"\n",
    "    enst_id = header_info.split()[0][1:]  # Remove '>' and extract ENST ID\n",
    "    \n",
    "    # Check if the extracted ID starts with 'ENST'\n",
    "    if enst_id.startswith('ENST'):\n",
    "        return enst_id\n",
    "    else:\n",
    "        print(f\"Invalid ENST ID in header: {header_info}\")\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_enst_dict(cds_dict):\n",
    "    \"\"\"\n",
    "    Create a dictionary mapping gene_symbols to their corresponding ENST IDs.\n",
    "    \"\"\"\n",
    "    enst_dict = {}\n",
    "    for gene, info in cds_dict.items():\n",
    "        enst_id = extract_enstID(info['cds_info'])\n",
    "        if enst_id:\n",
    "            enst_dict[gene] = enst_id  # Map gene name to ENST ID\n",
    "    return enst_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_cdna_dict(cdna_filepath):\n",
    "    \"\"\"\n",
    "    Build a dictionary with ENST IDs and cDNA seq\n",
    "    \"\"\"\n",
    "    cdna_dict = {}\n",
    "    current_enst_id = None\n",
    "    current_sequence = []\n",
    "\n",
    "    with open(cdna_filepath, 'r') as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                # If we have an ENST ID and a current sequence, store it in the dictionary\n",
    "                if current_enst_id is not None:\n",
    "                    cdna_dict[current_enst_id] = ''.join(current_sequence)\n",
    "\n",
    "                # Extract the new ENST ID from the current header\n",
    "                current_enst_id = extract_enstID(line)\n",
    "                current_sequence = []  # Reset sequence for the new capture\n",
    "            elif current_enst_id is not None:\n",
    "                current_sequence.append(line)\n",
    "\n",
    "        # Store the last entry if present\n",
    "        if current_enst_id is not None:\n",
    "            cdna_dict[current_enst_id] = ''.join(current_sequence)\n",
    "\n",
    "    return cdna_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_cdna_to_cds(cds_infodict, enst_dict, cdna_dict):\n",
    "    \"\"\"\n",
    "    Add cDNA sequences from cdna_dict to the cds_dict as a new item to each gene's entry in cds_dict.\n",
    "    \"\"\"\n",
    "    genes_without_cDNA = []\n",
    "\n",
    "    for gene_symbol, info in cds_infodict.items():\n",
    "        # Get the ENST ID for the current gene symbol\n",
    "        enst_id = enst_dict.get(gene_symbol)\n",
    "        \n",
    "        if enst_id:\n",
    "            # Check if the ENST ID exists in the cDNA dictionary\n",
    "            cdna_seq = cdna_dict.get(enst_id)\n",
    "            if cdna_seq:\n",
    "                # Add the cDNA sequence to the cds_dict entry\n",
    "                info['cdna_seq'] = cdna_seq\n",
    "            else:\n",
    "                print(f\"cDNA sequence for ENST ID {enst_id} not found.\")\n",
    "                # remove cases where no cDNA was found for ENST-ID\n",
    "                genes_without_cDNA.append(gene_symbol)\n",
    "        else:\n",
    "            print(f\"ENST ID for gene {gene_symbol} not found in enst_dict.\")\n",
    "\n",
    "    # Remove the marked genes from cds_dict\n",
    "    for gene_symbol in genes_without_cDNA:\n",
    "        del cds_infodict[gene_symbol]\n",
    "\n",
    "    return cds_infodict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n\\nBuild dictionarry with gene_symbol:ENST_ID\")\n",
    "enst_dict = create_enst_dict(cds_dict)\n",
    "\n",
    "print(\"\\n\\nBuild cDNA dictionary with ENST_ID:cDNA\")\n",
    "cdna_dict = build_cdna_dict(GRCh38_cDNA_fasta)\n",
    "\n",
    "print(\"\\n\\nFor ENST_ID with CDS, grep the cDNA seq\")\n",
    "gene_cdna_dict = {}\n",
    "n_ENST_notfound=0\n",
    "for gene, enst_id in enst_dict.items():\n",
    "    if enst_id in cdna_dict:\n",
    "        gene_cdna_dict[gene] = cdna_dict[enst_id]\n",
    "    else:\n",
    "        print(f\"ENST ID {enst_id} for gene {gene} not found in the cDNA file.\")\n",
    "        n_ENST_notfound+=1\n",
    "\n",
    "pct_noENSTID = n_ENST_notfound / (len(cds_dict.keys()) /100)\n",
    "pct_noENSTID = str(round(pct_noENSTID,2))\n",
    "\n",
    "print(f\"\\n\\nWe are working with {len(cdna_dict.keys())} cDNA sequences.\")\n",
    "print(f\"From {len(cds_dict.keys())} genes, {n_ENST_notfound} don't have a CDS with matched cDNA ENST-IDs ({pct_noENSTID}%).\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\nWe will drop all mitochondrial genes as theire translaction is more complex and follows different rules than nuclear genes (e.g. often no UTRs)\")\n",
    "# Dropping all genes marked with \"MT-\"\n",
    "cds_dict_noMT = {k: v for k, v in cds_dict.items() if not k.startswith('MT-')}\n",
    "print(f\"By dropping MT genes, we went from\",len(cds_dict.keys()),\"to\",len(cds_dict_noMT.keys()),\"number of cds', dropping\",len(cds_dict.keys())-len(cds_dict_noMT.keys()),\"genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18277"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(enst_dict.keys()) # 18290\n",
    "len(cdna_dict.keys()) # 206723 cDNAs\n",
    "len(cds_dict_noMT.keys()) # 18277 CDS matched with cDNA and ENST, removing mitochondrials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n\\nAdd the cDNA as a new item to the cds_dict\")\n",
    "complete_dict = add_cdna_to_cds(cds_dict_noMT, enst_dict, cdna_dict)\n",
    "print(f\"\\n\\nWe continue with {len(complete_dict.keys())} genes.\") #17369 genes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3: Mutate STOP codon\n",
    "**Step 3.1: Function *trim_cdna_to_overlap()* will keep only the 3'UTR of the cDNA (seq after end of CDS).**\n",
    "For all CDS with valid sequences and matched cDNA:\n",
    "- Drop the 5'UTR from the cDNAseq (keeping overlap + downstream cDNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def append_3UTR(sequences_dict, similarity_threshold=0.98):\n",
    "    \"\"\"\n",
    "    Find the overlapping sequence from the Stop of the cds to the end of the cdna_seq (3'UTR),\n",
    "    tolerating small differences between sequences.\n",
    "    Returns dictionary with the same genes as keys but adds a new item values are the overlapped sequences\n",
    "     \"\"\"\n",
    "    original_stop_codons = []\n",
    "\n",
    "    def find_stop_codon(sequence):\n",
    "        \"\"\"Find the first stop codon (TAA, TAG, or TGA) in the cds.\n",
    "        Return the position of the first stop codon found. \"\"\"\n",
    "        stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "        \n",
    "        # Find the first ATG\n",
    "        start_pos = sequence.find('ATG')\n",
    "        if start_pos == -1:\n",
    "            return -1,-1, ''\n",
    "        # Read sequence in triplets starting from ATG\n",
    "        pos = start_pos\n",
    "        while pos + 3 <= len(sequence):\n",
    "            codon = sequence[pos:pos + 3]\n",
    "            if codon in stop_codons:\n",
    "                return start_pos, pos, codon\n",
    "            pos += 3  # Move to next codon\n",
    "            \n",
    "        # if no stop codon found\n",
    "        return -1, -1, ''\n",
    "        \n",
    "\n",
    "    def calculate_similarity(seq1, seq2):\n",
    "        \"\"\"Calculate the ratio of matching nucleotides between two sequences.\"\"\"\n",
    "        if len(seq1) == 0:\n",
    "            return 0\n",
    "        if len(seq1) != len(seq2):\n",
    "            return 0\n",
    "        matches = sum(n1 == n2 for n1, n2 in zip(seq1, seq2))\n",
    "        return matches / len(seq1)\n",
    "    \n",
    "    def find_best_start_position(cds_start, cdna_seq):\n",
    "        \"\"\"Find the best matching position using a sliding window.\"\"\"\n",
    "        best_score = 0\n",
    "        best_pos = -1\n",
    "        \n",
    "        # Use the start of CDS sequence as a reference window\n",
    "        if len(cds_start) < 20 :\n",
    "            window_size = len(cds_start)\n",
    "        else:\n",
    "            window_size = 20\n",
    "        cds_window = cds_start[:window_size]\n",
    "\n",
    "\n",
    "        # Slide the window through cdna_seq\n",
    "        for i in range(len(cdna_seq) - window_size + 1):\n",
    "            cdna_window = cdna_seq[i:i + window_size]\n",
    "            score = calculate_similarity(cds_window, cdna_window)\n",
    "            \n",
    "            if score > best_score:\n",
    "                best_score = score\n",
    "                best_pos = i\n",
    "                \n",
    "            # Early exit if we find a very good match\n",
    "            if score > 0.95:\n",
    "                break\n",
    "                \n",
    "        return best_pos, best_score\n",
    "\n",
    "\n",
    "    trimmed_cDNA_dict = {}\n",
    "\n",
    "    for gene_id, info in sequences_dict.items():\n",
    "        # check if both seq exist\n",
    "        if 'cds_sequence' not in info or 'cdna_seq' not in info:\n",
    "            raise KeyError(f\"Missing required sequence keys for {gene_id}\")\n",
    "   \n",
    "        # Handle case-insensitive matching and leading N's\n",
    "        cds_seq = info['cds_sequence'].strip('Nn').upper()\n",
    "        cdna_seq = info['cdna_seq'].strip('Nn').upper()\n",
    "        \n",
    "        # Find where cds_seq starts in cdna_seq\n",
    "        cds_start_pos, similarity = find_best_start_position(cds_seq, cdna_seq)\n",
    "        \n",
    "        if cds_start_pos == -1 or similarity < similarity_threshold:\n",
    "            print(   f\"No sufficient match found for {gene_id}. \"\n",
    "                                f\"Best similarity: {similarity:.2f}\" )\n",
    "\n",
    "        # Find the stop codon in the overlapping cds region\n",
    "        aligned_seq = cdna_seq[cds_start_pos:] \n",
    "        start_pos, stop_pos, stop_codon = find_stop_codon(aligned_seq)\n",
    "\n",
    "        original_stop_codons.append(stop_codon)\n",
    "\n",
    "        # cut cds from start pos to end pos\n",
    "        cds_cut = aligned_seq[start_pos : stop_pos+3]\n",
    "\n",
    "        # Extract sequence from stop codon to end (including stop codon)\n",
    "        trimmed_cdna_seq = aligned_seq[stop_pos + 3:]\n",
    "\n",
    "        trimmed_cDNA_dict[gene_id] = {\n",
    "            'header' : info.get('cds_info'),  \n",
    "            'CDS' : cds_cut, \n",
    "            'CDS_uncut' : cds_seq,\n",
    "            'cDNA' : cdna_seq,\n",
    "            '3UTR' : trimmed_cdna_seq  \n",
    "        }\n",
    "    return trimmed_cDNA_dict,original_stop_codons\n",
    "\n",
    "print(\"\\n\\nAppend 3'UTR from cDNA to cds.\")\n",
    "complete_dict_3UTR,original_stop_codons = append_3UTR(complete_dict)\n",
    "# 18468, no cases removed due to cDNA-CDS unmatched pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"We continue working with\",len(complete_dict_3UTR.keys()),\"transcripts.\") # 17356"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'header': '>ENST00000409321.5 cds chromosome:GRCh38:2:231707684:231713108:1 gene:ENSG00000187514.17 gene_biotype:protein_coding transcript_biotype:protein_coding gene_symbol:PTMA description:prothymosin alpha [Source:HGNC Symbol;Acc:HGNC:9623]',\n",
       " 'CDS': 'ATGAATCCGTGTGTGCTTGGGACGGTGCCCGCTTGTCGACTGACTTGGCCGACGGACGCCGGACCTGATGACGCAGCCGTAGACACCAGCTCCGAAATCACCACCAAGGACTTAAAGGAGAAGAAGGAAGTTGTGGAAGAGGCAGAAAATGGAAGAGACGCCCCTGCTAACGGGAATGCTAATGAGGAAAATGGGGAGCAGGAGGCTGACAATGAGGTAGACGAAGAAGAGGAAGAAGGTGGGGAGGAAGAGGAGGAGGAAGAAGAAGGTGATGGTGAGGAAGAGGATGGAGATGAAGATGAGGAAGCTGAGTCAGCTACGGGCAAGCGGGCAGCTGAAGATGATGAGGATGACGATGTCGATACCAAGAAGCAGAAGACCGACGAGGATGACTAG',\n",
       " 'CDS_uncut': 'ATGAATCCGTGTGTGCTTGGGACGGTGCCCGCTTGTCGACTGACTTGGCCGACGGACGCCGGACCTGATGACGCAGCCGTAGACACCAGCTCCGAAATCACCACCAAGGACTTAAAGGAGAAGAAGGAAGTTGTGGAAGAGGCAGAAAATGGAAGAGACGCCCCTGCTAACGGGAATGCTAATGAGGAAAATGGGGAGCAGGAGGCTGACAATGAGGTAGACGAAGAAGAGGAAGAAGGTGGGGAGGAAGAGGAGGAGGAAGAAGAAGGTGATGGTGAGGAAGAGGATGGAGATGAAGATGAGGAAGCTGAGTCAGCTACGGGCAAGCGGGCAGCTGAAGATGATGAGGATGACGATGTCGATACCAAGAAGCAGAAGACCGACGAGGATGACTAG',\n",
       " 'cDNA': 'GGGACCGCTATGCCGTGCGGGGCCCGCGGGCTGTCCCACTCGAGGGCTGGAGCCGGCATCGCGCATCGAGTCCGCTTGATACATGAATCCGTGTGTGCTTGGGACGGTGCCCGCTTGTCGACTGACTTGGCCGACGGACGCCGGACCTGATGACGCAGCCGTAGACACCAGCTCCGAAATCACCACCAAGGACTTAAAGGAGAAGAAGGAAGTTGTGGAAGAGGCAGAAAATGGAAGAGACGCCCCTGCTAACGGGAATGCTAATGAGGAAAATGGGGAGCAGGAGGCTGACAATGAGGTAGACGAAGAAGAGGAAGAAGGTGGGGAGGAAGAGGAGGAGGAAGAAGAAGGTGATGGTGAGGAAGAGGATGGAGATGAAGATGAGGAAGCTGAGTCAGCTACGGGCAAGCGGGCAGCTGAAGATGATGAGGATGACGATGTCGATACCAAGAAGCAGAAGACCGACGAGGATGACTAGACAGCAAAAAAGGAAAAGTTAAACTAAAAAAAAAAAGGCCGCCGTGACCTATTCACCCTCCACTTCCCGTCTCAGAATCTAAACGTGGTCACCTTCGAGTAGAGAGGCCCGCCCGCCCACCGTGGGCAGTGCCACCCGCAGATGACACGCGCTCTCCACCACCCAACCCAAACCATGAGAATTTGCAACAGGGGAGGAAAAAAGAACCAAAACTTCCAAGGCCCTGCTTTTTTTCTTAAAAGTACTTTAAAAAGGAA',\n",
       " '3UTR': 'ACAGCAAAAAAGGAAAAGTTAAACTAAAAAAAAAAAGGCCGCCGTGACCTATTCACCCTCCACTTCCCGTCTCAGAATCTAAACGTGGTCACCTTCGAGTAGAGAGGCCCGCCCGCCCACCGTGGGCAGTGCCACCCGCAGATGACACGCGCTCTCCACCACCCAACCCAAACCATGAGAATTTGCAACAGGGGAGGAAAAAAGAACCAAAACTTCCAAGGCCCTGCTTTTTTTCTTAAAAGTACTTTAAAAAGGAA'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_dict_3UTR['PTMA']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 4: Getting seq of peptide extensions\n",
    "4.1 Create all possible mutations of cds\n",
    "\n",
    "\n",
    "4.2 Find next ds STOP\n",
    "4.2 translate to next ds STOP, if no dsSTop discard"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 4: Getting seq of peptide extensions\n",
    "4.1 Get downstream STOP codon in 3'UTR\n",
    "Since we can be sure that all cds seq in our dictionary are in the correct reading frame,\n",
    "we will just take the last 9 amino acids (8*3+3=27 nucleotides) of each seq\n",
    " (we controled)\n",
    "- passing 3codon window to find downstream STOP\n",
    "- if STOP found, cut 3'UTR to STOP\n",
    "- if no STOP in 3'UTR, delete entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_ds_STOP_in3UTR(trimmed_dict):\n",
    "    \"\"\"\n",
    "    Scan 3'UTR in frames of 3 nucleotides to look for a ds STOP.\n",
    "    If a ds STOP is found, trim 3'UTR to that position.\n",
    "    If no in-frame STOP is found, entry is dropped from dictionary.\n",
    "    \n",
    "    Returns a new dictionary with the 3'UTR trimmed.\n",
    "    \"\"\"\n",
    "    trimmed_UTR_dict = {}\n",
    "    no_dsSTOP_found = []\n",
    "    counter = 0\n",
    "    stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "    ds_STOPcodon = []\n",
    "\n",
    "    for gene_symbol, info in trimmed_dict.items():\n",
    "        utr_3_seq = info.get('3UTR')\n",
    "\n",
    "        # Slide through the 3' UTR in frames of 3 to find a downstream STOP codon\n",
    "        stop_found = False\n",
    "        for i in range(0, len(utr_3_seq), 3):\n",
    "            codon = utr_3_seq[i:i+3]\n",
    "            if codon in stop_codons:\n",
    "                stop_found = True\n",
    "                # store new stop for statistics\n",
    "                ds_STOPcodon.append(codon)\n",
    "                # Trim the 3' UTR up to and including this STOP codon\n",
    "                trimmed_UTR_seq = utr_3_seq[:i + 3]  # Keep up to STOP codon\n",
    "                # Create a new entry with the trimmed cdna_seq\n",
    "                trimmed_UTR_dict[gene_symbol] = {\n",
    "                    'header': info.get('header'),  \n",
    "                    'CDS': info.get('CDS'),  \n",
    "                    'cDNA': info.get('cDNA'),\n",
    "                    '3UTR': info.get('3UTR'),\n",
    "                    'UTR_trimmed': trimmed_UTR_seq  \n",
    "                }\n",
    "                break\n",
    "\n",
    "        # If no downstream STOP codon is found, print a warning and remove gene from dict\n",
    "        if not stop_found:\n",
    "            no_dsSTOP_found.append(gene_symbol)\n",
    "            counter += 1\n",
    "            print(f\"No downstream in-frame STOP codon found for gene {gene_symbol}. Entry will be removed.\")\n",
    "\n",
    "    genes_before = len(trimmed_dict.keys())\n",
    "    pctt_genes_noSTOP = counter / ( genes_before /100) \n",
    "    pctt_genes_noSTOP = str(round(pctt_genes_noSTOP,2))\n",
    "    print(f\"Out of {len(trimmed_dict.keys())}, we removed {counter} genes due to missing dsSTOP ({pctt_genes_noSTOP}%).\")\n",
    "\n",
    "    return trimmed_UTR_dict,ds_STOPcodon\n",
    "\n",
    "print(\"\\n\\nDrop 3'UTR after the first ds STROP codon.\")\n",
    "trimmed_UTR_dict, new_ds_STOPcodons = find_ds_STOP_in3UTR(complete_dict_3UTR)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare if the distributin of STOP codons changes between the cds and the ds STOPs in the 3'UTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import chi2_contingency\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "stop_codons = ['TGA', 'TAG', 'TAA']\n",
    "\n",
    "# Convert Counters to lists of observed frequencies\n",
    "original_counter = Counter(original_stop_codons)\n",
    "new_counter = Counter(new_ds_STOPcodons)\n",
    "\n",
    "# Create lists of frequencies\n",
    "obs_values = [original_counter.get(codon, 0) for codon in stop_codons]\n",
    "new_values = [new_counter.get(codon, 0) for codon in stop_codons]\n",
    "\n",
    "# Perform chi-square test\n",
    "chi2_stat, p_value = chi2_contingency([obs_values, new_values])[0:2]\n",
    "\n",
    "print(\"Stop codon occurence in unodified transcripts:\")\n",
    "print(Counter(original_stop_codons))\n",
    "print(\"Down-stream stop codon occurence after stoploss:\")\n",
    "print(Counter(new_ds_STOPcodons))\n",
    "print(f\"Chi-square statistic: {chi2_stat}\")\n",
    "print(f\"p-value: {p_value}\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot stop codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "# Get counts\n",
    "original_counts = [Counter(original_stop_codons).get(codon, 0) for codon in stop_codons]\n",
    "new_counts = [Counter(new_ds_STOPcodons).get(codon, 0) for codon in stop_codons]\n",
    "x = np.arange(len(stop_codons))\n",
    "width = 0.35\n",
    "fig, ax = plt.subplots(figsize=(5, 6))\n",
    "# Create bars\n",
    "ax.bar(x - width/2, original_counts, width, label='CDS', color='#333333')\n",
    "ax.bar(x + width/2, new_counts, width, label=\"3'UTR\", color='#a4bad0')\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Occurrences')\n",
    "ax.set_title('Stop Codon Comparison', pad=20)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(stop_codons)\n",
    "ax.legend()\n",
    "p_value_text = f'Chi-square p-value: {p_value:.4f}'\n",
    "if p_value < 0.05:\n",
    "    p_value_text += ' *'\n",
    "ax.text(0.5, 0.95, p_value_text, transform=ax.transAxes, \n",
    "        horizontalalignment='center')\n",
    "plt.tight_layout()\n",
    "plt.savefig(wdir+\"STOPcodon_occurence.png\")\n",
    "plt.close()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Stacked barplot of frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "# Function to count the proportions of each stop codon in a list\n",
    "def count_proportions(lst):\n",
    "    stop_codons = ['TGA', 'TAA', 'TAG']\n",
    "    count = {codon: lst.count(codon) for codon in stop_codons}\n",
    "    total = sum(count.values())\n",
    "    proportions = {codon: count[codon] / total for codon in stop_codons}\n",
    "    return proportions\n",
    "\n",
    "original_props = count_proportions(original_stop_codons)\n",
    "ds_props = count_proportions(new_ds_STOPcodons)\n",
    "\n",
    "# Create a DataFrame for easy plotting\n",
    "df_prop = pd.DataFrame({\n",
    "    'TGA': [original_props['TGA'], ds_props['TGA']],\n",
    "    'TAA': [original_props['TAA'], ds_props['TAA']],\n",
    "    'TAG': [original_props['TAG'], ds_props['TAG']],\n",
    "}, index=['CDS', \"3'UTR\"])\n",
    "\n",
    "observed = [list(original_props.values()), list(ds_props.values())]\n",
    "# Perform the Chi-squared test\n",
    "chi2_stat, p_value, dof, expected = stats.chi2_contingency(observed)\n",
    "\n",
    "# Plot the stacked bar chart\n",
    "ax = df_prop.plot(kind='bar', stacked=True, figsize=(5, 6), color=['#a3bac0', '#49636b', '#1d282b'])\n",
    "ax.set_title('Stop Codon Proportion Comparison', pad=40)\n",
    "plt.ylabel(\"Proportion\")\n",
    "plt.xticks(rotation=0)\n",
    "plt.legend(title='Stop Codons')\n",
    "plt.tight_layout()\n",
    "plt.figtext(0.35, 0.9, f'Chi-squared p-value = {p_value:.4f}', fontsize=12, color='black')\n",
    "plt.savefig(wdir+\"STOPcodon_proportions.png\")\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TGA': 0.49913564596058546,\n",
       " 'TAA': 0.2774576466520687,\n",
       " 'TAG': 0.22340670738734586}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_props"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_symbol = \"PTMA\"\n",
    "entry = trimmed_UTR_dict[gene_symbol]\n",
    "print(f\"Gene: {gene_symbol}\")\n",
    "print(f\"Header: {entry['header']}\")\n",
    "print(f\"CDS: {entry['CDS']}\")\n",
    "print(f\"cDNA: {entry['cDNA']}\")\n",
    "print(f\"3UTR: {entry['3UTR']}\")\n",
    "print(f\"UTR: {entry['UTR_trimmed']}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Put together the new peptide sequences with all possible mutations of the STOP codon\n",
    "- Get last 8 codons from cds (including STOP in last position)\n",
    "- Depending on which last STOP codon (TAA, TGA, TAG), mutated the last position\n",
    "- Join mutated sequence with the trimmed 3'UTR tail\n",
    "- translate the sequneces and store the new AA in fasta (us as headers gene name and STOP codon mutation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_mutated_sequences_with_utr(input_dict):\n",
    "    \"\"\"\n",
    "    Apply utations to the STOP codon, and append the trimmed UTR sequence:\n",
    "    Mutated sequence: base_seq_without_stop + mutated_STOP + utr_3_seq\n",
    "\n",
    "    Returns a new dictionary with mutated sequences for each gene.\n",
    "    \"\"\"\n",
    "    nonsilentSNV_dict ={ \n",
    "        \"TAG\" : [\"GAG\",\"TGG\",\"AAG\",\"CAG\",\"TCG\",\"TAC\",\"TTG\"],    # TAT codes as TTG for Tyr\n",
    "        \"TAA\" : [\"GAA\",\"AAA\",\"CAA\",\"TCA\",\"TAC\",\"TTA\"],          # TAU codes as TAC for Tyr \n",
    "        \"TGA\" : [\"GGA\",\"TGG\",\"AGA\",\"TCA\",\"TGC\",\"TTA\"],          # CGA codes as AGA for Arg, TGT codes as TGC for Cys\n",
    "    }\n",
    "    # Set counters for the STOP codons\n",
    "    n_TAG=0\n",
    "    n_TAA=0\n",
    "    n_TGA=0\n",
    "\n",
    "    mutated_dict = {}\n",
    "\n",
    "    for gene_symbol, info in input_dict.items():\n",
    "        cds_seq = info.get('CDS')\n",
    "        utr_3_seq = info.get('UTR_trimmed')\n",
    "\n",
    "        if cds_seq and utr_3_seq:\n",
    "            # Extract the last 9 codons (27 nucleotides), the last 3 should be STOP\n",
    "            last_8_codon_seq = cds_seq[-27:] \n",
    "            stop_codon = last_8_codon_seq[-3:] \n",
    "\n",
    "            # Ensure the STOP codon is TAG, TAA or TGA\n",
    "            if stop_codon in nonsilentSNV_dict:\n",
    "                # counter for STOP codons\n",
    "                if   stop_codon == \"TAG\": n_TAG += 1\n",
    "                elif stop_codon == \"TAA\": n_TAA += 1\n",
    "                elif stop_codon == \"TGA\": n_TGA += 1\n",
    "\n",
    "                mutated_STOPcodons = nonsilentSNV_dict[stop_codon] # get the mutated codons\n",
    "                base_seq_without_stop = last_8_codon_seq[:-3]  # Drop STOP codon \n",
    "\n",
    "                # Store cds and the mutated sequences for this gene\n",
    "                not_mutated_seq = []\n",
    "                mutated_seqs_full = []\n",
    "                mutated_seqs_8 = []\n",
    "\n",
    "                # Generate mutated sequences\n",
    "                for mutated_codon in mutated_STOPcodons:\n",
    "                    # Full mutated sequence: CDS + mutated_STOP + utr_3_seq\n",
    "                    mutated_seq_full = cds_seq[:-3] + mutated_codon + utr_3_seq\n",
    "                    # Last 8 canonical Aa + mutated_STOP + utr_3_seq\n",
    "                    mutated_seq_8 = base_seq_without_stop + mutated_codon + utr_3_seq\n",
    "                    # Make list with all mutated sequences for the gene\n",
    "                    not_mutated_seq.append(cds_seq)\n",
    "                    mutated_seqs_full.append(mutated_seq_full)\n",
    "                    mutated_seqs_8.append(mutated_seq_8)\n",
    "\n",
    "                mutated_dict[gene_symbol] = {\n",
    "                    'header': info.get('header'),  \n",
    "                    'cds_seq' : not_mutated_seq,\n",
    "                    'mutated_seqs_full': mutated_seqs_full , \n",
    "                    'mutated_seqs_8': mutated_seqs_8  \n",
    "                }\n",
    "            else:\n",
    "                print(f\"Warning: Unrecognized STOP codon '{stop_codon}' for gene {gene_symbol}. Skipping mutations.\")\n",
    "        else:\n",
    "            print(f\"Warning: Missing CDS or 3' UTR sequence for gene {gene_symbol}.\")\n",
    "\n",
    "    print(f\"STOP codon counts: TAG = {n_TAG}, TAA = {n_TAA}, TGA = {n_TGA}\")\n",
    "\n",
    "    return mutated_dict\n",
    "\n",
    "print(\"\\n\\nCreate mutated sequences from CDS+mutated STOP+3'UTR to next STOP.\")\n",
    "mutated_seq_dict = generate_mutated_sequences_with_utr(trimmed_UTR_dict)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### STEP 3.4: Translate the mutated seq to peptide format and store in fasta outfile\n",
    "- Translate the mutated CDS sequences into peptides\n",
    "- Make unique IDs for each new seq: [gene_symbol].X_[ENST_ID]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_header(header, gene_symbol, i):\n",
    "    \"\"\"\n",
    "    Updates the header so that the first string after '>' is 'gene_symbol+i', and the rest of the header remains unchanged.\n",
    "    \"\"\"\n",
    "    # Split the header by the first space to separate the ENST ID from the rest of the header\n",
    "    parts = header.split(' ', 1)\n",
    "    \n",
    "    # Create the new header by replacing the ENST ID with 'gene_symbol+i'\n",
    "    new_header = f\">{gene_symbol}.{i} {parts[1]}\"  # Reconstruct the header\n",
    "    \n",
    "    return new_header\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate seq from cds nuc to AA pep\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "def translate_mutated_cds_to_pep(nuc_dict):\n",
    "    \"\"\"\n",
    "    Translates the mutated sequences from mutated_seq_dict to peptide sequences and stores them in a new dictionary.\n",
    "    A new dictionary with gene_symbol as keys and the translated peptide sequence and header as values.\n",
    "    \"\"\"\n",
    "    peptide_dict = {}\n",
    "\n",
    "    genes_inframe_stp = []\n",
    "\n",
    "    for gene_symbol, info in nuc_dict.items():\n",
    "        cds_seq_list = info.get('cds_seq')  # Get the full mutated CDS nucleotide sequence\n",
    "        nuc_seq_full_list = info.get('mutated_seqs_full')  # Get the full mutated CDS nucleotide sequence\n",
    "        nuc_seq_8_list = info.get('mutated_seqs_8')  # Get the shortened CDS nucleotide sequence\n",
    "        header = info.get('header')     # Get the header information\n",
    "\n",
    "        if len(nuc_seq_full_list) == 0:\n",
    "            print(f\"No mutated sequences found. Skip {gene_symbol}\")\n",
    "            break\n",
    "        \n",
    "        for i in range(len(nuc_seq_full_list)):\n",
    "            # Translate not mutated CDS\n",
    "            cds_seq = cds_seq_list[i]\n",
    "            cds_seq_obj = Seq(cds_seq) # using Biopython's Seq object724594\n",
    "            if str(cds_seq[:3]).upper() != \"ATG\":\n",
    "                print(gene_symbol)\n",
    "                print(f\"ERROR: No Start codon in not mutated cds seq:\",cds_seq)\n",
    "            else:\n",
    "                # check if we have an extra in-frame STOP codon\n",
    "                n = len(cds_seq)\n",
    "                stp_counter = 0\n",
    "                for j in range(0, n - n % 3, 3):\n",
    "                    codon = cds_seq[j : j + 3]\n",
    "                    if codon in stop_codons:\n",
    "                        stp_counter += 1\n",
    "                \n",
    "                if stp_counter >= 2:\n",
    "                    genes_inframe_stp.append(gene_symbol)\n",
    "                else:\n",
    "                    cds_peptide = str(cds_seq_obj.translate(to_stop=True))  # Translate from ATG until the first STOP codon\n",
    "\n",
    "            # Translate full mutated CDS\n",
    "            nuc_seq_full = nuc_seq_full_list[i]\n",
    "            nuc_seq_full_obj = Seq(nuc_seq_full) # using Biopython's Seq object724594\n",
    "            if str(nuc_seq_full[:3]).upper() != \"ATG\":\n",
    "                print(gene_symbol)\n",
    "                print(f\"ERROR: No Start codon in seq:\",nuc_seq_full)\n",
    "            else:\n",
    "                # check if we have an extra in-frame STOP codon\n",
    "                n = len(nuc_seq_full)\n",
    "                stp_counter = 0\n",
    "                for j in range(0, n - n % 3, 3):\n",
    "                    codon = nuc_seq_full[j : j + 3]\n",
    "                    if codon in stop_codons:\n",
    "                        stp_counter += 1\n",
    "                \n",
    "                if stp_counter >= 2:\n",
    "                    genes_inframe_stp.append(gene_symbol)\n",
    "                else:\n",
    "                    full_peptide_seq = str(nuc_seq_full_obj.translate(to_stop=True))  # Translate from ATG until the first STOP codon\n",
    "\n",
    "            # Translate shortened mutated CDS\n",
    "            nuc_seq_8 = nuc_seq_8_list[i]\n",
    "            nuc_seq_obj = Seq(nuc_seq_8)\n",
    "            peptide_8_seq = str(nuc_seq_obj.translate(to_stop=True))  # Translate until the first STOP codon\n",
    "            # comment: we cannot use 'cds=True' in the translate function, as we cut of the ATG together with \n",
    "            # the beginning of the cds and only kept its last 8 codons (incl the previous STOP codon) \n",
    " \n",
    "            # make unique ID for mutated seq\n",
    "            updated_header = update_header(header, gene_symbol, i)\n",
    "            \n",
    "            \n",
    "            # Store the peptide sequence and the header in the new dictionary\n",
    "            peptide_dict[updated_header] = {\n",
    "                # not mutated cds\n",
    "                'pep_cds' : cds_peptide, \n",
    "                'cds_seq' : cds_seq,\n",
    "                # cds + elongations\n",
    "                'pep_AA_full': full_peptide_seq,\n",
    "                'mutated_cds_full': nuc_seq_full,\n",
    "                # last 8 AA of cds + elongations\n",
    "                'pep_AA_8': peptide_8_seq,\n",
    "                'mutated_cds_8': nuc_seq_8,\n",
    "                # pep elongation only\n",
    "                'elongation' : peptide_8_seq[9:]\n",
    "            }\n",
    "\n",
    "    # remove gene with in-frame stop in cds\n",
    "    print(f\"For\",len(set(genes_inframe_stp)),\"genes, the cds contains an in-frame stop and will be dropped.\")\n",
    "    \n",
    "    return peptide_dict\n",
    "\n",
    "print(\"\\n\\nTranslate mutated sequences to amino acid.\")\n",
    "pep_cds_dict = translate_mutated_cds_to_pep(mutated_seq_dict)\n",
    "print(f\"We have mutated the CDS from {len(mutated_seq_dict.keys())} genes.\")\n",
    "print(f\"This resulted in {len(pep_cds_dict.keys())} peptide sequences from STOPloss mutations.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'>PKD1L2.0 cds chromosome:GRCh38:16:81170289:81181329:-1 gene:ENSG00000166473.17 gene_biotype:polymorphic_pseudogene transcript_biotype:protein_coding gene_symbol:PKD1L2 description:polycystin 1 like 2 (gene/pseudogene) [Source:HGNC Symbol;Acc:HGNC:21715]': {'pep_cds': 'MGEDSPVAMFSWYLDNTPTEQAEPLLDACRLRGFWPRSLTLLQSNTSTLLLNSSFLQSRGEVIRIRATALTRHAYGEDTYVISTVPPREVPACTIAPEEGTVLTSFAIFCNASTALGPLEFCFCLESGSCLHCGPEPALPSVYLPLGEENNDFVLTVVISATNRAGDTQQTQAMAKVALGDTCVEDVAFQAAVSEKIPTALQGEGGPEQLLQLAKAVSSMLNQEHESQGSGQSLSIDVRQKVREHVLGSLSAVTTGLEDVQRVQELAEVLREVTCRSKELTPSAQGSCMGDSWEGAPPAAHVSHAR',\n",
       "  'cds_seq': 'ATGGGGGAGGACTCTCCAGTGGCTATGTTCAGCTGGTATTTGGACAACACCCCAACAGAGCAGGCTGAGCCCCTCCTGGATGCCTGCAGACTCAGAGGATTTTGGCCAAGGTCCTTAACCCTCCTCCAGAGCAACACCTCCACGTTGCTGTTGAACAGCTCGTTTCTGCAGTCCCGGGGAGAGGTCATCCGAATCAGAGCCACAGCACTGACCAGGCATGCCTATGGGGAGGACACCTATGTGATCAGCACTGTGCCTCCCCGTGAGGTGCCTGCCTGCACTATTGCCCCAGAGGAGGGCACCGTTCTGACGAGCTTTGCCATCTTCTGCAACGCCTCCACAGCCCTGGGACCCCTGGAGTTCTGCTTCTGTCTGGAATCAGGTTCCTGCCTACACTGTGGCCCTGAACCTGCCCTCCCATCAGTGTATCTGCCACTTGGAGAGGAGAACAATGACTTTGTGCTGACAGTAGTTATTTCTGCCACCAATCGTGCAGGGGACACGCAGCAGACCCAGGCCATGGCTAAGGTGGCACTCGGAGACACATGTGTTGAGGATGTAGCATTCCAGGCTGCCGTGTCAGAGAAAATCCCCACAGCTCTGCAAGGCGAGGGTGGCCCCGAGCAGCTCCTCCAGCTGGCCAAGGCTGTGTCCTCCATGCTGAACCAAGAGCATGAAAGCCAGGGCTCAGGACAGTCACTGAGCATAGACGTCAGACAGAAGGTCAGAGAGCATGTGCTGGGATCACTGTCTGCAGTCACCACCGGCTTGGAGGACGTGCAGAGGGTGCAGGAGCTGGCCGAGGTGCTGAGAGAGGTGACCTGCCGGAGTAAGGAACTCACACCCTCGGCCCAGGGGTCCTGCATGGGCGATTCATGGGAAGGTGCCCCTCCTGCTGCCCATGTATCTCACGCTAGGTGA',\n",
       "  'pep_AA_full': 'MGEDSPVAMFSWYLDNTPTEQAEPLLDACRLRGFWPRSLTLLQSNTSTLLLNSSFLQSRGEVIRIRATALTRHAYGEDTYVISTVPPREVPACTIAPEEGTVLTSFAIFCNASTALGPLEFCFCLESGSCLHCGPEPALPSVYLPLGEENNDFVLTVVISATNRAGDTQQTQAMAKVALGDTCVEDVAFQAAVSEKIPTALQGEGGPEQLLQLAKAVSSMLNQEHESQGSGQSLSIDVRQKVREHVLGSLSAVTTGLEDVQRVQELAEVLREVTCRSKELTPSAQGSCMGDSWEGAPPAAHVSHARGEGLFAQTLTPASAGEQVEGVTESH',\n",
       "  'mutated_cds_full': 'ATGGGGGAGGACTCTCCAGTGGCTATGTTCAGCTGGTATTTGGACAACACCCCAACAGAGCAGGCTGAGCCCCTCCTGGATGCCTGCAGACTCAGAGGATTTTGGCCAAGGTCCTTAACCCTCCTCCAGAGCAACACCTCCACGTTGCTGTTGAACAGCTCGTTTCTGCAGTCCCGGGGAGAGGTCATCCGAATCAGAGCCACAGCACTGACCAGGCATGCCTATGGGGAGGACACCTATGTGATCAGCACTGTGCCTCCCCGTGAGGTGCCTGCCTGCACTATTGCCCCAGAGGAGGGCACCGTTCTGACGAGCTTTGCCATCTTCTGCAACGCCTCCACAGCCCTGGGACCCCTGGAGTTCTGCTTCTGTCTGGAATCAGGTTCCTGCCTACACTGTGGCCCTGAACCTGCCCTCCCATCAGTGTATCTGCCACTTGGAGAGGAGAACAATGACTTTGTGCTGACAGTAGTTATTTCTGCCACCAATCGTGCAGGGGACACGCAGCAGACCCAGGCCATGGCTAAGGTGGCACTCGGAGACACATGTGTTGAGGATGTAGCATTCCAGGCTGCCGTGTCAGAGAAAATCCCCACAGCTCTGCAAGGCGAGGGTGGCCCCGAGCAGCTCCTCCAGCTGGCCAAGGCTGTGTCCTCCATGCTGAACCAAGAGCATGAAAGCCAGGGCTCAGGACAGTCACTGAGCATAGACGTCAGACAGAAGGTCAGAGAGCATGTGCTGGGATCACTGTCTGCAGTCACCACCGGCTTGGAGGACGTGCAGAGGGTGCAGGAGCTGGCCGAGGTGCTGAGAGAGGTGACCTGCCGGAGTAAGGAACTCACACCCTCGGCCCAGGGGTCCTGCATGGGCGATTCATGGGAAGGTGCCCCTCCTGCTGCCCATGTATCTCACGCTAGGGGAGAGGGCCTGTTTGCCCAGACTCTCACTCCTGCATCTGCTGGTGAGCAAGTTGAGGGAGTAACTGAATCTCATTAA',\n",
       "  'pep_AA_8': 'AAHVSHARGEGLFAQTLTPASAGEQVEGVTESH',\n",
       "  'mutated_cds_8': 'GCTGCCCATGTATCTCACGCTAGGGGAGAGGGCCTGTTTGCCCAGACTCTCACTCCTGCATCTGCTGGTGAGCAAGTTGAGGGAGTAACTGAATCTCATTAA',\n",
       "  'elongation': 'EGLFAQTLTPASAGEQVEGVTESH'}}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict(list(pep_cds_dict.items())[0: 1]) "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4: Write Fasta Files\n",
    "Function *write_fasta_files()* creates 2 output files:\n",
    "        - peptide seq in amino acids\n",
    "        - mutated and elongated CDS in nucleotides\n",
    "\n",
    "The final two fasta files contain the exact same sequneces of the mutated sequences\n",
    "in amino acid (pep) and nucleotide (CDS) format.\n",
    "\n",
    "**Mutated sequence: [8 last codons of CDS] + [mutated STOP codon] + [3'UTR to downstream STOP codon]**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fasta_files(seq_dict,out_path):\n",
    "    \"\"\"\n",
    "    Write two FASTA files: one for peptide sequences and one for mutated CDS sequences.\n",
    "    \"\"\"\n",
    "    # the last 8 canonical AA + elongations\n",
    "    cds_fasta_path = outpath+'CDS_elongation_seq.fasta'\n",
    "    pep_fasta_path = outpath+'peptide_elongations.fasta'\n",
    "    # the whole cds + elonagtions\n",
    "    pep_fasta_full_path = outpath+'peptide_whole_transcript_elongations.fasta'\n",
    "    cds_fasta_full_path = outpath+'CDS_whole_transcript_elongation_seq.fasta'\n",
    "    # only the cds (not mutated)\n",
    "    pep_fasta_notmutated_path = outpath+'CDS_peptide_notmutated.fasta'\n",
    "    cds_fasta_notmutated_path = outpath+'CDS_transcript_notmutated.fasta'\n",
    "\n",
    "    with open(pep_fasta_path, 'w') as pep_fasta, open(pep_fasta_full_path, 'w') as pep_full_fasta, open(pep_fasta_notmutated_path, 'w') as pep_fasta_notmutated, open(cds_fasta_path, 'w') as cds_fasta_8, open(cds_fasta_full_path, 'w') as cds_fasta_full,  open(cds_fasta_notmutated_path, 'w') as cds_fasta_notmutated:\n",
    "        for header, sequences in seq_dict.items():\n",
    "            pep_seq_full = sequences.get('pep_AA_full')\n",
    "            pep_seq_8 = sequences.get('pep_AA_8')  # Get peptide sequence\n",
    "            mutated_cds_seq_full = sequences.get('mutated_cds_full')  # Get mutated CDS sequence\n",
    "            mutated_cds_seq_8 = sequences.get('mutated_cds_8')  # Get mutated CDS sequence\n",
    "\n",
    "            # Write the peptide sequence to the peptide FASTA file\n",
    "            if pep_seq_full:\n",
    "                pep_full_fasta.write(f\"{header}\\n\")\n",
    "                pep_full_fasta.write(f\"{pep_seq_full}\\n\")\n",
    "            \n",
    "            if pep_seq_8:\n",
    "                pep_fasta.write(f\"{header}\\n\")\n",
    "                pep_fasta.write(f\"{pep_seq_8}\\n\")\n",
    "\n",
    "            # Write the mutated CDS sequence to the CDS FASTA file\n",
    "            if mutated_cds_seq_full:\n",
    "                cds_fasta_full.write(f\"{header}\\n\")\n",
    "                cds_fasta_full.write(f\"{mutated_cds_seq_full}\\n\")\n",
    "\n",
    "            if mutated_cds_seq_8:\n",
    "                cds_fasta_8.write(f\"{header}\\n\")\n",
    "                cds_fasta_8.write(f\"{mutated_cds_seq_8}\\n\")\n",
    "\n",
    "            # Write the CDS sequence (without mutations) to the CDS FASTA file\n",
    "            cds_seq = sequences.get('cds_seq') \n",
    "            if cds_seq:\n",
    "                cds_fasta_notmutated.write(f\"{header}\\n\")\n",
    "                cds_fasta_notmutated.write(f\"{cds_seq}\\n\")\n",
    "\n",
    "            cds_pep = sequences.get('pep_cds') \n",
    "            if cds_pep:\n",
    "                pep_fasta_notmutated.write(f\"{header}\\n\")\n",
    "                pep_fasta_notmutated.write(f\"{cds_pep}\\n\")\n",
    "\n",
    "print(\"\\n\\nSave amino acids and mutated nucleotide sequences to fasta files.\")\n",
    "write_fasta_files(pep_cds_dict, outpath )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'>PKD1L2': 'EGLFAQTLTPASAGEQVEGVTESH',\n",
       " '>INTS3': 'GPAFPIPPPAGLPSPSW',\n",
       " '>KIR2DL3': 'SKVVSCP',\n",
       " '>KIR2DL4': 'RRESPS',\n",
       " '>KIR2DL1': 'APQSGLEGVF',\n",
       " '>KIR3DL1': 'APQSGLEDIF',\n",
       " '>KIR2DS3': 'LDHCVFTQRKITPPSQRPKTPPTDSSMYIELPNAESRSKAVFCPRAPQSGLEGIF',\n",
       " '>KIR2DL2': 'APQSGLEGVF',\n",
       " '>PCMTD2': 'VSCLKGGNRKSRLLSVKFVLPVCC',\n",
       " '>KIR3DP1': 'FCSICLK'}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# file with Gene Symbol and only elongation (without mutated STOP)\n",
    "def consolidate_gene_elongation(input_dict):\n",
    "    # Create a new dictionary to store the consolidated results\n",
    "    consolidated_dict = {}\n",
    "    \n",
    "    # Iterate through the input dictionary\n",
    "    for full_key, value_dict in input_dict.items():\n",
    "        # Extract the base gene name (removing version number)\n",
    "        base_gene = full_key.split(' ')[0].rsplit('.', 1)[0]\n",
    "        \n",
    "        # If this base gene is not in the consolidated dict, add it\n",
    "        if base_gene not in consolidated_dict:\n",
    "            # Check if 'elongation' exists in the current value dictionary\n",
    "            if 'elongation' in value_dict:\n",
    "                consolidated_dict[base_gene] = value_dict['elongation']\n",
    "    \n",
    "    # print to fasta file\n",
    "    pep_elongation_fasta_path = outpath+'Elongation_only_pep.fasta'\n",
    "    with open(pep_elongation_fasta_path, 'w') as pep_elongation_fasta:\n",
    "        for header, sequence in consolidated_dict.items():\n",
    "            pep_elongation_fasta.write(f\"{header}\\n\")\n",
    "            pep_elongation_fasta.write(f\"{sequence}\\n\")\n",
    "\n",
    "    return consolidated_dict\n",
    "\n",
    "gene_elong_dict = consolidate_gene_elongation(pep_cds_dict)\n",
    "len(gene_elong_dict) # 16073 genes\n",
    "dict(list(gene_elong_dict.items())[0:10]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n\\nVersion:\")\n",
    "print(sys.version)\n",
    "\n",
    "sys.stdout = old_stdout\n",
    "\n",
    "log_file.close()\n",
    "\n",
    "print(\"All done. Gute Nacht.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
